smart-contract-feature-development.md

```diff
@@ -0,0 +1,94 @@
+are in a state such that the custom gas token feature needs to be ported to the `OptimismPortal2` before the `OptimismPortal` and the
+`OptimismPortal2` can be merged. This may be a good solution for security critical code as we do not rely on the correctness
+of solidity inheritance and all of the logic lives in a single contract, making it easier to audit.
+
+# Proposed Solution
+
+Below includes general guidelines for how to develop solidity feature contracts that are meant to be merged into mainline at
+some point in the future.
+
+### Composition through Inheritance
+
+Solidity allows contacts to [inherit](https://docs.soliditylang.org/en/v0.8.17/contracts.html#inheritance) other contracts.
+The functionality of the contract can then be extended or overridden.
+For in development features that modify core contracts already running in production,
+a new contract should be created that inherits the core contract and all new functionality should be added to the new contract.
+The deploy script should be able to parse some sort of config and optionally deploy the feature enabled version of the contract.
+
+### Handling Inheritance
+
+Inherit the contract that needs new functionality. It is very easy to add net new functions. If certain functions need to be modified,
+then the original contract may need to have the `virtual` keyword added in places. The `virtual` keyword should only be added in
+places where it must be added and nowhere in addition.
+
+### Handling Semver
+
+The parent contract should update its `version` getter to be a real function. The child contract should implement a `version` getter like so:
+
+```solidity
+function version() public override returns (string) {
+  return string.concat(super.version(), "+feature-name");
+}
+```
+
+The `feature-name` should be some identifier that uniquely identifies the feature. It could also be a hardfork name if the release is tied to a hard fork.
+
+### Handling Storage
+
+Handling storage becomes complex and there is no solution that handles all cases well. We generally want to ensure that the contracts
+are forwards compatible, so if `solc` is managing storage, the base contract should have spacers when we know that a feature contract
+is scheduled to be merged soon.
+
+Another solution is to use unstructured storage, meaning not letting `solc` manage storage and instead using custom storage slots.
+Assuming that digests and unique preimages are used for the storage slot keys, then this can work well for ensuring there are no
+storage slot collisions. This works well if there is no need for `solc` to manage the storage layout.
+
+If there are multiple feature contracts based on the same base contract being worked on in parallel, then there is complexity around
+layout management and it is up to the teams working on features to coordinate with each other if they risk mangling storage layouts.
+Using unstructured storage greatly reduces the coordination overhead here.
```

Comment:
Another idea: When writing your feature contract, always add storage variables to the base contract, and everything else (constants, events, initializers, core logic, etc) is added to the feature contract. This way, when two feature contracts are in being worked on in parallel, the rule is "only append to the base contract storage", which should reduce the coordination required and reduce the chance of storage layout issues

Reply:
I do like this idea altho it adds risk around code that may be shipped to prod being changed due to changes in experimental features. If the experimental feature changes, then it would result in changes in storage layout in the base contract. I suppose that could be ok, if its always empty